File: packages/tx402-python/src/tx402/protocol.py
"""Strict x402 v2 challenge decoding and tx402 normalization (SPEC §6.2)."""

from __future__ import annotations

import base64
import binascii
import hashlib
import json
import re
from datetime import datetime, timezone
from typing import Any, Final
from urllib.parse import urlsplit

from x402.http.utils import decode_payment_required_header

from tx402.canonical_json import canonicalize_json
from tx402.errors import (
    InvalidPaymentRequiredError,
    Tx402ErrorContext,
    UnsupportedProtocolError,
)

MAX_PAYMENT_REQUIRED_BYTES: Final = 64 * 1024
MAX_PAYMENT_REQUIRED_DEPTH: Final = 16
MAX_PAYMENT_REQUIREMENTS: Final = 32
_BASE64: Final = re.compile(r"^[A-Za-z0-9+/]*={0,2}$")
_CAIP2: Final = re.compile(r"^[a-z0-9-]{3,8}:[A-Za-z0-9-]{1,48}$")
_AMOUNT: Final = re.compile(r"^[1-9][0-9]*$")


class _PreflightError(ValueError):
    def __init__(self, reason: str) -> None:
        super().__init__(reason)
        self.reason = reason


def _pairs(pairs: list[tuple[str, Any]]) -> dict[str, Any]:
    result: dict[str, Any] = {}
    for key, value in pairs:
        if key in result:
            raise _PreflightError("duplicate-json-key")
        result[key] = value
    return result


def _check_depth(value: Any, depth: int = 1) -> None:
    if isinstance(value, (dict, list)):
        if depth > MAX_PAYMENT_REQUIRED_DEPTH:
            raise _PreflightError("json-depth-exceeded")
        children = value.values() if isinstance(value, dict) else value
        for child in children:
            _check_depth(child, depth + 1)


def _invalid(
    reason: str,
    context: Tx402ErrorContext,
    schema_path: str = "/",
    cause: BaseException | None = None,
) -> InvalidPaymentRequiredError:
    return InvalidPaymentRequiredError(
        f"Invalid PAYMENT-REQUIRED challenge: {reason}",
        context=context,
        details={"reason": reason, "schemaPath": schema_path},
        cause=cause,
    )


def _origin(url: str) -> tuple[str, str, int] | None:
    parsed = urlsplit(url)
    if parsed.scheme not in {"http", "https"} or parsed.hostname is None:
        return None
    port = parsed.port or (443 if parsed.scheme == "https" else 80)
    return parsed.scheme, parsed.hostname.lower(), port


def _digest(value: str) -> str:
    return f"sha256:{hashlib.sha256(value.encode('utf-8')).hexdigest()}"


def decode_payment_required(
    header: str | None,
    *,
    request_url: str,
    request_method: str,
    request_id: str,
    clock_epoch_ms: int,
) -> dict[str, Any]:
    """Decode, bound, validate, and normalize one PAYMENT-REQUIRED header."""
    context = Tx402ErrorContext(request_id=request_id, phase="parse")
    if not header:
        raise _invalid("missing-header", context)
    if _BASE64.fullmatch(header) is None or len(header) % 4 != 0:
        raise _invalid("invalid-base64", context)

    try:
        raw = base64.b64decode(header, validate=True)
    except (binascii.Error, ValueError) as error:
        raise _invalid("invalid-base64", context, cause=error) from error
    if len(raw) > MAX_PAYMENT_REQUIRED_BYTES:
        raise _invalid("header-too-large", context)

    try:
        text = raw.decode("utf-8")
        document = json.loads(text, object_pairs_hook=_pairs)
        _check_depth(document)
    except _PreflightError as error:
        raise _invalid(error.reason, context, cause=error) from error
    except (UnicodeDecodeError, json.JSONDecodeError) as error:
        raise _invalid("invalid-json", context, cause=error) from error

    if not isinstance(document, dict):
        raise _invalid("upstream-schema-invalid", context)
    version = document.get("x402Version")
    if version != 2:
        raise UnsupportedProtocolError(
            "Unsupported x402 protocol version",
            context=context,
            details={
                "observedVersion": version,
                "supportedVersions": [2],
                "reason": "unsupported-protocol-version",
            },
        )

    accepts = document.get("accepts")
    if not isinstance(accepts, list) or not 1 <= len(accepts) <= MAX_PAYMENT_REQUIREMENTS:
        raise _invalid("requirements-count-out-of-range", context, "/accepts")
    resource = document.get("resource")
    if not isinstance(resource, dict) or not isinstance(resource.get("url"), str):
        raise _invalid("upstream-schema-invalid", context, "/resource")
    declared_url = resource["url"]
    if _origin(declared_url) is None or _origin(declared_url) != _origin(request_url):
        raise _invalid("resource-origin-mismatch", context, "/resource/url")

    for index, requirement in enumerate(accepts):
        if not isinstance(requirement, dict):
            raise _invalid("upstream-schema-invalid", context, f"/accepts/{index}")
        amount = requirement.get("amount")
        if not isinstance(amount, str) or _AMOUNT.fullmatch(amount) is None:
            raise _invalid("amount-not-atomic-integer", context, f"/accepts/{index}/amount")

    try:
        upstream = decode_payment_required_header(header)
        upstream_document = upstream.model_dump(by_alias=True, exclude_none=True)
    except (ValueError, TypeError) as error:
        raise _invalid("upstream-schema-invalid", context, cause=error) from error

    normalized_requirements: list[dict[str, Any]] = []
    for index, requirement in enumerate(upstream_document["accepts"]):
        if (
            not isinstance(requirement.get("scheme"), str)
            or not requirement["scheme"]
            or not isinstance(requirement.get("network"), str)
            or _CAIP2.fullmatch(requirement["network"]) is None
            or not isinstance(requirement.get("asset"), str)
            or not requirement["asset"]
            or not isinstance(requirement.get("payTo"), str)
            or not requirement["payTo"]
            or not isinstance(requirement.get("maxTimeoutSeconds"), int)
            or not 1 <= requirement["maxTimeoutSeconds"] <= 86_400
            or not isinstance(requirement.get("extra"), dict)
        ):
            raise _invalid("upstream-schema-invalid", context, f"/accepts/{index}")
        normalized_requirements.append(
            {
                "index": index,
                "scheme": requirement["scheme"],
                "network": requirement["network"],
                "asset": requirement["asset"],
                "amountAtomic": requirement["amount"],
                "payTo": requirement["payTo"],
                "maxTimeoutSeconds": requirement["maxTimeoutSeconds"],
                "extra": requirement["extra"],
                "rawHash": _digest(canonicalize_json(requirement)),
            }
        )

    received_at = datetime.fromtimestamp(clock_epoch_ms / 1000, tz=timezone.utc)
    normalized: dict[str, Any] = {
        "protocolVersion": 2,
        "resource": {"url": declared_url, "method": request_method.upper()},
        "requirements": normalized_requirements,
        "receivedAt": received_at.isoformat(timespec="milliseconds").replace("+00:00", "Z"),
        "headerHash": _digest(header),
    }
    if isinstance(upstream_document.get("error"), str):
        normalized["error"] = upstream_document["error"]
    return normalized

File: packages/tx402/src/core/protocol.ts
/** Strict x402 v2 challenge decoding and tx402 normalization (SPEC §6.2). */

import { createHash } from "node:crypto";

import { decodePaymentRequiredHeader } from "@x402/core/http";

import { canonicalizeJson } from "./canonical-json.js";
import {
  InvalidPaymentRequiredError,
  UnsupportedProtocolError,
  type Tx402ErrorContext,
} from "./errors.js";

export const MAX_PAYMENT_REQUIRED_BYTES = 64 * 1024;
export const MAX_PAYMENT_REQUIRED_DEPTH = 16;
export const MAX_PAYMENT_REQUIREMENTS = 32;

export interface NormalizedPaymentRequirement {
  readonly index: number;
  readonly scheme: string;
  readonly network: string;
  readonly asset: string;
  readonly amountAtomic: string;
  readonly payTo: string;
  readonly maxTimeoutSeconds: number;
  readonly extra: Readonly<Record<string, unknown>>;
  readonly rawHash: string;
}

export interface NormalizedPaymentRequired {
  readonly protocolVersion: 2;
  readonly resource: Readonly<{ url: string; method: string }>;
  readonly requirements: readonly NormalizedPaymentRequirement[];
  readonly receivedAt: string;
  readonly headerHash: string;
  readonly error?: string;
}

export interface DecodePaymentRequiredOptions {
  readonly requestUrl: string | URL;
  readonly requestMethod: string;
  readonly requestId: string;
  readonly clockEpochMs?: number;
  readonly allowInsecureLocalhost?: boolean;
}

type UpstreamRequirement = {
  scheme: string;
  network: string;
  asset: string;
  amount: string;
  payTo: string;
  maxTimeoutSeconds: number;
  extra: Record<string, unknown>;
};

type UpstreamPaymentRequired = {
  x402Version: number;
  resource: { url: string };
  accepts: UpstreamRequirement[];
  error?: string;
};

class JsonPreflightError extends Error {
  constructor(
    readonly reason: "duplicate-json-key" | "json-depth-exceeded" | "invalid-json",
  ) {
    super(reason);
  }
}

/**
 * Validates JSON structure before `JSON.parse`, which otherwise silently keeps the last
 * duplicate object member. This deliberately parses only grammar and keys; upstream still
 * owns conversion of the envelope into its protocol object (ADR-010 decision 6).
 */
function preflightJson(text: string): void {
  let offset = 0;

  const whitespace = (): void => {
    while (/\s/u.test(text[offset] ?? "")) offset += 1;
  };

  const string = (): string => {
    const start = offset;
    if (text[offset] !== '"') throw new JsonPreflightError("invalid-json");
    offset += 1;
    while (offset < text.length) {
      const character = text[offset];
      if (character === '"') {
        offset += 1;
        try {
          return JSON.parse(text.slice(start, offset)) as string;
        } catch {
          throw new JsonPreflightError("invalid-json");
        }
      }
      if (character === "\\") {
        offset += 2;
      } else {
        if (character === undefined || character.charCodeAt(0) < 0x20) {
          throw new JsonPreflightError("invalid-json");
        }
        offset += 1;
      }
    }
    throw new JsonPreflightError("invalid-json");
  };

  const value = (depth: number): void => {
    whitespace();
    const character = text[offset];
    if (character === "{" || character === "[") {
      if (depth > MAX_PAYMENT_REQUIRED_DEPTH) {
        throw new JsonPreflightError("json-depth-exceeded");
      }
      if (character === "{") object(depth);
      else array(depth);
      return;
    }
    if (character === '"') {
      string();
      return;
    }
    const rest = text.slice(offset);
    const token = /^(?:true|false|null|-?(?:0|[1-9]\d*)(?:\.\d+)?(?:[eE][+-]?\d+)?)/u.exec(
      rest,
    )?.[0];
    if (!token) throw new JsonPreflightError("invalid-json");
    offset += token.length;
  };

  const object = (depth: number): void => {
    offset += 1;
    whitespace();
    const keys = new Set<string>();
    if (text[offset] === "}") {
      offset += 1;
      return;
    }
    while (true) {
      whitespace();
      const key = string();
      if (keys.has(key)) throw new JsonPreflightError("duplicate-json-key");
      keys.add(key);
      whitespace();
      if (text[offset] !== ":") throw new JsonPreflightError("invalid-json");
      offset += 1;
      value(depth + 1);
      whitespace();
      if (text[offset] === "}") {
        offset += 1;
        return;
      }
      if (text[offset] !== ",") throw new JsonPreflightError("invalid-json");
      offset += 1;
    }
  };

  const array = (depth: number): void => {
    offset += 1;
    whitespace();
    if (text[offset] === "]") {
      offset += 1;
      return;
    }
    while (true) {
      value(depth + 1);
      whitespace();
      if (text[offset] === "]") {
        offset += 1;
        return;
      }
      if (text[offset] !== ",") throw new JsonPreflightError("invalid-json");
      offset += 1;
    }
  };

  value(1);
  whitespace();
  if (offset !== text.length) throw new JsonPreflightError("invalid-json");
}

function digest(value: string): string {
  return `sha256:${createHash("sha256").update(value, "utf8").digest("hex")}`;
}

function invalid(
  reason: string,
  context: Tx402ErrorContext,
  schemaPath = "/",
  cause?: unknown,
): InvalidPaymentRequiredError {
  return new InvalidPaymentRequiredError(`Invalid PAYMENT-REQUIRED challenge: ${reason}`, {
    context,
    details: { reason, schemaPath },
    ...(cause === undefined ? {} : { cause }),
  });
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

function isLocalhost(hostname: string): boolean {
  const normalized = hostname.toLowerCase().replace(/^\[|\]$/gu, "");
  return normalized === "localhost" || normalized === "127.0.0.1" || normalized === "::1";
}

function validateRequirement(
  value: unknown,
  index: number,
  context: Tx402ErrorContext,
): UpstreamRequirement {
  if (!isRecord(value))
    throw invalid("upstream-schema-invalid", context, `/accepts/${index}`);
  const { scheme, network, asset, amount, payTo, maxTimeoutSeconds, extra } = value;
  if (
    typeof scheme !== "string" ||
    scheme.length === 0 ||
    scheme.length > 64 ||
    typeof network !== "string" ||
    !/^[a-z0-9-]{3,8}:[A-Za-z0-9-]{1,48}$/u.test(network) ||
    typeof asset !== "string" ||
    asset.length === 0 ||
    asset.length > 128 ||
    typeof payTo !== "string" ||
    payTo.length === 0 ||
    payTo.length > 128 ||
    !Number.isInteger(maxTimeoutSeconds) ||
    (maxTimeoutSeconds as number) < 1 ||
    (maxTimeoutSeconds as number) > 86_400 ||
    !isRecord(extra)
  ) {
    throw invalid("upstream-schema-invalid", context, `/accepts/${index}`);
  }
  if (typeof amount !== "string" || !/^[1-9][0-9]*$/u.test(amount)) {
    throw invalid("amount-not-atomic-integer", context, `/accepts/${index}/amount`);
  }
  return {
    scheme,
    network,
    asset,
    amount,
    payTo,
    maxTimeoutSeconds: maxTimeoutSeconds as number,
    extra,
  };
}

/** Decode, bound, validate, and normalize one raw PAYMENT-REQUIRED header. */
export function decodePaymentRequired(
  header: string | null | undefined,
  options: DecodePaymentRequiredOptions,
): NormalizedPaymentRequired {
  const context: Tx402ErrorContext = { requestId: options.requestId, phase: "parse" };
  if (header === null || header === undefined || header.length === 0) {
    throw invalid("missing-header", context);
  }
  if (!/^[A-Za-z0-9+/]*={0,2}$/u.test(header) || header.length % 4 !== 0) {
    throw invalid("invalid-base64", context);
  }

  let bytes: Uint8Array;
  let jsonText: string;
  try {
    bytes = Buffer.from(header, "base64");
    if (bytes.byteLength > MAX_PAYMENT_REQUIRED_BYTES)
      throw invalid("header-too-large", context);
    jsonText = new TextDecoder("utf-8", { fatal: true }).decode(bytes);
  } catch (error) {
    if (error instanceof InvalidPaymentRequiredError) throw error;
    throw invalid("invalid-base64", context, "/", error);
  }

  try {
    preflightJson(jsonText);
  } catch (error) {
    const reason = error instanceof JsonPreflightError ? error.reason : "invalid-json";
    throw invalid(reason, context, "/", error);
  }

  let decoded: unknown;
  try {
    decoded = decodePaymentRequiredHeader(header);
  } catch (error) {
    throw invalid("invalid-json", context, "/", error);
  }
  if (!isRecord(decoded)) throw invalid("upstream-schema-invalid", context);
  if (decoded.x402Version !== 2) {
    throw new UnsupportedProtocolError("Unsupported x402 protocol version", {
      context,
      details: {
        observedVersion: decoded.x402Version,
        supportedVersions: [2],
        reason: "unsupported-protocol-version",
      },
    });
  }
  if (!isRecord(decoded.resource) || typeof decoded.resource.url !== "string") {
    throw invalid("upstream-schema-invalid", context, "/resource");
  }
  if (
    !Array.isArray(decoded.accepts) ||
    decoded.accepts.length < 1 ||
    decoded.accepts.length > MAX_PAYMENT_REQUIREMENTS
  ) {
    throw invalid("requirements-count-out-of-range", context, "/accepts");
  }

  let declaredUrl: URL;
  let requestUrl: URL;
  try {
    declaredUrl = new URL(decoded.resource.url);
    requestUrl = new URL(options.requestUrl);
  } catch (error) {
    throw invalid("resource-url-invalid", context, "/resource/url", error);
  }
  const secure = declaredUrl.protocol === "https:";
  const permittedLocalHttp =
    options.allowInsecureLocalhost === true &&
    declaredUrl.protocol === "http:" &&
    requestUrl.protocol === "http:" &&
    isLocalhost(declaredUrl.hostname) &&
    isLocalhost(requestUrl.hostname);
  if ((!secure && !permittedLocalHttp) || declaredUrl.origin !== requestUrl.origin) {
    throw invalid("resource-origin-mismatch", context, "/resource/url");
  }

  const upstream = decoded as unknown as UpstreamPaymentRequired;
  const requirements = upstream.accepts.map((entry, index) => {
    const requirement = validateRequirement(entry, index, context);
    return Object.freeze({
      index,
      scheme: requirement.scheme,
      network: requirement.network,
      asset: requirement.asset,
      amountAtomic: requirement.amount,
      payTo: requirement.payTo,
      maxTimeoutSeconds: requirement.maxTimeoutSeconds,
      extra: Object.freeze({ ...requirement.extra }),
      rawHash: (() => {
        try {
          return digest(canonicalizeJson(requirement));
        } catch (error) {
          throw invalid("upstream-schema-invalid", context, `/accepts/${index}`, error);
        }
      })(),
    });
  });

  const normalized = {
    protocolVersion: 2 as const,
    resource: Object.freeze({
      url: decoded.resource.url,
      method: options.requestMethod.toUpperCase(),
    }),
    requirements: Object.freeze(requirements),
    receivedAt: new Date(options.clockEpochMs ?? Date.now()).toISOString(),
    headerHash: digest(header),
    ...(typeof decoded.error === "string" ? { error: decoded.error } : {}),
  };
  return Object.freeze(normalized);
}
